### ISSUE TITLE
Path autosuggestion can hang on network access

### ISSUE THREAD

**Issue by RSully** [fish-shell/fish-shell#3206]
The only similar issue I found is #2771. When using a network drive (SMB in my case) fish will attempt ~~autocompletion~~ autosuggestion but sometimes get stuck waiting for network access to complete for various reasons. During this time I am unable to continue typing the path. 

## Reproduction steps

1. Launch fish
2. Connect to a slow or unreliable SMB file server
3. Try to `cd` into that mounted path

## Expected results

Fish should do suggestion requests asynchronously. Alternatively, maybe let me blacklist `/Volumes` and/or `/mnt`.

## Actual results

I have to wait until the network request finishes or times out.

**Comment by zanchey** [comment]
To clarify, do you need to press tab for this to occur? That makes the difference between a completion and an autosuggestion.

**Comment by krader1961** [comment]
Also, did you really mean to say "Try to cd into that mounted path"? The reason I ask is that doing the auto path completion asynchronously doesn't help if you actually type `cd /broken/mount/point`. So I'm curious what you expect fish to do in that scenario even if the auto completion is done in an async manner.

**Comment by RSully** [comment]
@zanchey no tab

@krader1961 sorry could you rephrase that?

**Comment by krader1961** [comment]
What do you expect fish to do if you type `cd /broken/smb/mount` (assuming auto-path completion didn't hang)? I'm asking because that's what you said you were doing when you noticed the auto-path completion hung the shell.

**Comment by RSully** [comment]
@zanchey thought I imagine the same issue exists for completion, maybe it is less of a UX problem

@krader1961 so if I have something mounted to `/mnt/unreliable` and I'm in `/mnt/unreliable/media` and start to type `cd j` (beginning to type `jimi`) during a time when my server (or, also likely, macOS client) begins to act up, fish will hang after the j. `jimi hendrix` may exist as a valid path, but the filesystem is being slow to respond or is going to timeout (in some cases when macOS is being really unreliable it'll eventually timeout and cause the network share to disconnect).

**Comment by krader1961** [comment]
I understand the issue with the auto suggestion. Let me try again. Pretend fish did not auto suggest a possible path and thus won't hang when you type `j`. You type `cd 'jimi hendrix'` then press [enter]. Since the remote file system is "slow to respond" what do you expect fish to do when it issues the `chdir("jimi hendrix")` system call?

**Comment by RSully** [comment]
In that case I don't think fish could do anything but wait synchronously, but I'd hope I could Ctrl+C it. What I hate is when I don't have control over my shell (i.e. can't kill whatever is happening), which is the case for with autosuggestions.

**Comment by krader1961** [comment]
Whether or not a file system related syscall is interruptible depends on the underlying file system type (e.g., local, SMB, NFS) and mount options (e.g., was the `intr` option used for an NFS mount). I just wanted to make sure you weren't expecting behavior that fish can't possibly provide (e.g., we can't do the `chdir()` call asynchronously). While we can in principal keep generating path auto suggestions from hanging the shell it may be prohibitively expensive in terms of code complexity.

**Comment by faho** [comment]
To reiterate what I said in #2771:

I think completions and suggestions should be stopped after a certain timeout, with suggestions having a much stricter one (<500ms?). These features are meant to save time, not add to it.

**Comment by RSully** [comment]
@faho sure, but making it async could handle that case as well, without losing the functionality. I'm not sure how suggestions work under-the-hood, but as I keep typing (in an async world) previous suggestion queries could become constrained so that they don't have to wait for an entire directory listing, for example. I do realize this probably adds to the code complexity as it stands now, or at the very least requires a refactoring to make it maintainable.

With completion it is a different story, though maybe there could be a way to allow users to cancel completions, rather than dying after an arbitrary timeout. In most cases I am using completion because I kind of know the filename I want, but not really, so if a completion were to timeout I'd just end up grepping the directory listing, wasting more time and using more resources.

**Comment by ridiculousfish** [comment]
fish's autosuggestions are supposed to execute in a background thread specifically so that they do not hang on slow mounts.

**Comment by alphapapa** [comment]
Guessing this should be tagged as a bug needing more info.

**Comment by ridiculousfish** [comment]
I still can't reproduce this even when I use my [test slow filesystem](https://github.com/ridiculousfish/sleepyfs). What I can reproduce is that even operations outside the slow filesystem can be slowed down, but this is due to how macOS works (not sure why, but definitely not something fish can do or avoid).

@RSully can you grab the backtrace as I described above? If not we'll have to close as cannot repro.

**Comment by RSully** [comment]
Yeah sorry been busy! I will try to reproduce soon though since I want to do it before I upgrade to 10.12 just incase things changed (I've had a few bugs marked as "try to test again on 10.12" so I'm thinking they redid some of their SMB support). Should *hopefully* have a chance in the next 1-2 weeks?

**Comment by ridiculousfish** [comment]
Ok!

**Comment by krader1961** [comment]
Ping. Another month has gone by and we still don't even know if the problem can be reproduced.

**Comment by RSully** [comment]
Alright so I started playing with it. When I manually "tab" to complete something, I can get fish to hang. I believe this may be expected, even if it is annoying. Perhaps the most annoying part is that the user can't cancel this action.

When I start typing a path to get it to auto-suggest it doesn't hang from what I've seen so far. Maybe this was a bug in an older version of fish.

**Comment by krader1961** [comment]
@ridiculousfish knows this portion of the fish code far better than I do. So I'll only speak as a former level three UNIX support engineer who did kernel crash analysis on a daily basis :smile: Completing a path is going to involve system calls like `stat()` and `access()`. If the underlying VFS (virtual file system) object that any portion of the path corresponds to is slow to respond to those syscalls then fish will be slow. And there is no way, given the UNIX APIs, to impose an upper limit on how long we're willing to wait before getting an answer.

Theoretically fish could be structured to perform all such syscalls in auxiliary threads. But that has its own problems. Since we don't know if the syscall will ever finish we would have to start a new thread for each such syscall. So if you have a major meltdown with your remote filesystems fish could easily spawn thousands of threads trying to do path completions. Obviously we can put a limit on the number of outstanding path completion threads (e.g., by using a thread pool). But now you've just moved the problem down the road a little ways and immensely complicated the code. And in the end the user is still likely to see weird behavior.

Finally, this sort of problem can even occur with a nominally "local" filesystem. Think about a SAN (storage area network) of disks accessed over Fiber Channel (FC) with multiple routes to each disk and multiple systems accessing each disk. I've seen failure modes that can result in small but noticeable delays in such configurations.

**Comment by faho** [comment]
>When I manually "tab" to complete something, I can get fish to hang. I believe this may be expected, even if it is annoying. Perhaps the most annoying part is that the user can't cancel this action.

In that case, this is a duplicate of #23.

**Comment by ridiculousfish** [comment]
Thanks for taking another look @RSully! Is this still with the remote SMB filesystem on macOS?
